src/robust_provider/subscription.rs

```diff
@@ -0,0 +1,1213 @@
+/// and periodic reconnection attempts to the primary provider.
+#[derive(Debug)]
+pub struct RobustSubscription<N: Network> {
+    subscription: Option<Subscription<N::HeaderResponse>>,
+    robust_provider: RobustProvider<N>,
+    last_reconnect_attempt: Option<Instant>,
+    consecutive_lags: usize,
+    current_fallback_index: Option<usize>,
+}
+
+impl<N: Network> RobustSubscription<N> {
+    /// Create a new [`RobustSubscription`]
+    pub(crate) fn new(
+        subscription: Subscription<N::HeaderResponse>,
+        robust_provider: RobustProvider<N>,
+    ) -> Self {
+        Self {
+            subscription: Some(subscription),
+            robust_provider,
+            last_reconnect_attempt: None,
+            consecutive_lags: 0,
+            current_fallback_index: None,
+        }
+    }
+
+    /// Receive the next item from the subscription with automatic failover.
+    ///
+    /// This method will:
+    /// * Attempt to receive from the current subscription
+    /// * Handle errors by switching to fallback providers
+    /// * Periodically attempt to reconnect to the primary provider
+    /// * Will switch to fallback providers if subscription timeout is exhausted
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if all providers have been exhausted and failed.
+    pub async fn recv(&mut self) -> Result<N::HeaderResponse, Error> {
+        let subscription_timeout = self.robust_provider.subscription_timeout;
+        while let Some(subscription) = &mut self.subscription {
+            let recv_result = timeout(subscription_timeout, subscription.recv()).await;
+            match recv_result {
+                Ok(recv_result) => match recv_result {
+                    Ok(header) => {
+                        if self.is_on_fallback() {
+                            self.try_reconnect_to_primary(false).await;
+                        }
+                        self.consecutive_lags = 0;
+                        return Ok(header);
+                    }
+                    Err(recv_error) => {
+                        self.process_recv_error(recv_error).await?;
+                    }
+                },
+                Err(elapsed_err) => {
+                    error!(
+                        timeout_secs = subscription_timeout.as_secs(),
+                        "Subscription timeout - no block received, switching provider"
+                    );
+
+                    self.switch_to_fallback(elapsed_err.into()).await?;
+                }
+            }
+        }
+        // No subscription available
+        Err(RpcError::Transport(TransportErrorKind::BackendGone).into())
+    }
+
+    /// Process subscription receive errors and handle failover
+    async fn process_recv_error(&mut self, recv_error: RecvError) -> Result<(), Error> {
+        match recv_error {
+            RecvError::Closed => {
+                error!("Subscription channel closed, switching provider");
```

Comment:
```suggestion
                error!("Provider closed the subscription channel, switching provider");
```
nit: explicit about this being an external error